File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    v = [0]*len(dungeons)
    lst = []
    def dfs(tmp,n):
        if n == len(dungeons):
            lst.append(tmp[:])
            return
        for i in range(len(dungeons)):
            if v[i]==0:
                v[i]=1
                tmp.append(dungeons[i])
                dfs(tmp,n+1)
                tmp.pop()
                v[i]=0
    
    dfs([],0)
    cntlst = []
    for dun in lst:
        nk = k
        cnt = 0
        for x,y in dun:
            if nk >= x:
                cnt+=1
                nk-=y
            else:
                break
        cntlst.append(cnt)
    return max(cntlst)

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int dfs(vector<bool> visited, int k, vector<vector<int>> dungeons){
    int max_dungeons_count = 0;
    
    for (int i = 0; i < visited.size(); i++){
        if (!visited[i] && k >= dungeons[i][0]){
            visited[i] = true;
            int current_dungeons = 1 + dfs(visited, k - dungeons[i][1], dungeons);
            max_dungeons_count = max(max_dungeons_count, current_dungeons);
            visited[i] = false;
        }
    }
    return max_dungeons_count;
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = 0;
    vector<bool> visited(dungeons.size(), false);
    return dfs(visited, k, dungeons);
    return answer;
}

// n와 m 문제 처럼 모든 경우를 다 가보기
// 80
// 80 20
// 50 40
// 30 10